0011-S-Saga-pattern-support-in-dapr-workflow.md

```diff
@@ -0,0 +1,803 @@
+
+```java
+public abstract class Workflow {
+  ......
+  public boolean isSagaEnabled() {
+    return this.getSagaOption() != null;
+  }
+  public SagaOption getSagaOption() {
+    return null;
+  }
+}
+```
+
+The isSagaEnabled and SagaOption are hard code in the Dapr workflow implementation class. So Dapr workflow can get back this information in the replay.
+
+The Saga object will be passed into WorkflowContext. So later workflow can use it by "ctx.registerCompensation()" / "ctx.compensate()" and provide built-in support.
+
+#### Task1: collect compensation information
+
+Compensation information is saved for each workflow instance execution. Suppose an activity (in the below example, UpdateInventoryActivity) is executed successfully, and it needs to do compensation when a workflow fails. In that case, the related information about the compensation activity (in the below example, UpdateInventoryCompensationActivity) is saved in the Saga object(in memory): 
+
+```java
+  public WorkflowStub create() {
+    return ctx -> {
+      ......
+        inventoryResult = ctx.callActivity(UpdateInventoryActivity.class.getName(),
+            inventoryRequest, InventoryResult.class).await();
+        if (!inventoryResult.isSuccess()) {
+					......
+          return;
+        }
+        // ** Task1: Collect the compensation information during workflow execution**
+        ctx.registerCompensation(UpdateInventoryCompensationActivity.class.getName(), inventoryRequest);
+      ......
+    }
+  }
+```
+
+Compensation will be performed later based on the saved compensation information if needed. To support retry in compensation activities, registerCompensation() method will have an override method to set its retry strategy:
+
+```java
+ctx.registerCompensation(
+              UpdateInventoryCompensationActivity.class.getName(), 
+              inventoryRequest);	
+ctx.registerCompensation(
+              UpdateInventoryCompensationActivity.class.getName(), 
+              inventoryRequest, taskOptions);
+```
+
+The customer must write a corresponding compensation activity for each activity to be compensated.
+
+```java
+public class ProcessPaymentActivity implements WorkflowActivity {......}
+public class ProcessPaymentCompensationActivity implements CompensatableWorkflowActivity {......}
+
+public class UpdateInventoryActivity implements WorkflowActivity {......}
+public class UpdateInventoryCompensationActivity implements CompensatableWorkflowActivity {......}
+```
+
+These compensation activities should also be registered like normal activities:
+
+```java
+    WorkflowRuntimeBuilder builder = new WorkflowRuntimeBuilder().registerWorkflow(OrderProcessingWorkflow.class);
+    builder.registerActivity(ReserveInventoryActivity.class);
+    builder.registerActivity(UpdateInventoryActivity.class);
+		......
+    builder.registerActivity(ProcessPaymentCompensationActivity.class);
+    builder.registerActivity(UpdateInventoryCompensationActivity.class);
+```
```

Comment:
Currently workflow has no concept of saga and compensation, so it will not call the Compensate() method on compensation, only the run() method.

But I think there is value in your suggestion: it reduces the number of activities that need to be registered, and it allows run() and compensate() to be placed in a single activity implementation class.

The changes needed workflow runtime to provide support, and the runtime would need to make a judgment here: if it's compensation it would execute the compensata() method instead of the run() method. The input for this judgment could be passed in by the dapr sdk, something like ctx.callCompensateActiviy(), or by adding a parameter to ctx.callActivity() to indicate that it's a compensation operation.

Additionally, two history records need to be saved in the activity history for this purpose: one for the run() method and one for the compensate() method.

@cgillum , help me to see how much this change costs and if it's worth doing.